src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static class Node<T> {
+        private T data;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(T data, Node<T> prev, Node<T> next) {
+            this.data = data;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value, last, null);
+        if (last != null) {
+            last.next = newNode;
+        }
+        last = newNode;
+        if (first == null) {
+            first = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> current = getNodeAtIndex(index);
+        Node<T> newNode = new Node<>(value, current.prev, current);
+        if (current.prev != null) {
+            current.prev.next = newNode;
+        } else {
+            first = newNode;
+        }
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeAtIndex(index);
+        return current.data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeAtIndex(index);
+        T oldValue = current.data;
+        current.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+
+        Node<T> current = getNodeAtIndex(index);
+
+        if (current.prev != null) {
+            current.prev.next = current.next;
+        } else {
+            first = current.next;
+        }
+
+        if (current.next != null) {
+            current.next.prev = current.prev;
+        } else {
+            last = current.prev;
+        }
```

Comment:
Did you check common mistakes : https://github.com/mate-academy/jv-linked-list/blob/master/checklist.md

I think that you can create some private method